### ISSUE TITLE
data_retention_time_in_days = 0 must be defined in snowflake_database from share

### ISSUE THREAD

**Issue by nicolamarangoni** [Snowflake-Labs/terraform-provider-snowflake#2483]
### Terraform CLI and Provider Versions

Provider version: 0.85.0

### Terraform Configuration

```terraform
resource "snowflake_database" "from_share" {
  for_each                    = var.databases_from_share
  
  name                        = upper(each.key)
  # data_retention_time_in_days = 0
  from_share = {
    provider = upper(each.value.from_share.provider)
    share    = upper(each.value.from_share.share)
  }
}
```

### Expected Behavior

After successfully creating the database from share I can add other resources and apply the changes.

### Actual Behavior

After successfully creating the database from share I cannot add other resources and apply the changes.
all successive applies try to set `data_retention_time_in_days = 1`:
Terraform will perform the following actions:
```
  # snowflake_database.from_share["my_share"] will be updated in-place
  ~ resource "snowflake_database" "from_share" {
      ~ data_retention_time_in_days = 0 -> 1
        id                          = "MY_SHARE"
        name                        = "MY_SHARE"
        # (2 unchanged attributes hidden)
    }
    ...
snowflake_database.from_share["my_share"]: Modifying... [id=MY_SHARE]
╷
│ Error: error updating database data retention time on MY_SHARE err = 003001 (42501): SQL access control error:
│ Insufficient privileges to operate on database 'MY_SHARE'
│ 
│   with snowflake_database.from_share["my_share"],
│   on snowflake_shares.tf line 1, in resource "snowflake_database" "from_share":
│    1: resource "snowflake_database" "from_share" {
│ 
```
**Adding data_retention_time_in_days = 0 to the database definition fixes this:
Terraform doesn't detect any changes in the database from share**

### Steps to Reproduce

1. `terraform apply` without `data_retention_time_in_days`
2. `terraform apply` another time: the second time terraform want to set `data_retention_time_in_days = 0 -> 1`

### How much impact is this issue causing?

Medium

**Comment by sfc-gh-jcieslak** [comment]
Hey @nicolamarangoni 👋 
This looks like a Snowflake error indicating you have insufficient privileges rather than a TF Provider issue. If you used other role than ACCOUNTADMIN to create/modify the resource then I assume there may be some complications [doc reference](https://docs.snowflake.com/en/user-guide/data-share-consumers#creating-a-database-from-a-share).

**Comment by nicolamarangoni** [comment]
in snowflake as ACCOUNTADMIN:
`alter database MY_SHARE set DATA_RETENTION_TIME_IN_DAYS = 1;`
Error:
`Insufficient privileges to operate on database 'MY_SHARE'`
This is correct because the DB is created from a share. I cannot change the retention of a DB from another account.
I also understand that you should not define a **data_retention_time_in_days** when creating a database from a share.
**The issue with the provider is that it tries to modify the retention of a database created from a share.**
It detects that the retention is 0 days for the existing database from the share and then want to set 1 day, what isn't defined anywhere in my deployment.
**What I expect is that retention is not concerned at all when creating databases from shares.**

**Comment by sfc-gh-jcieslak** [comment]
Ok, I didn't analyze it in-depth, but this looks like related to this issue - https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2356. So, the issue may be happening, because there's a default value set on `data_retention_time_in_days`. When it's not set in the config and remotely it's set to something different than 1, Terraform tries to set it back to 1 (because that's the default). I'm working on a change right now that will remove the default, so after a merge and release it should work as intended.

**Comment by sfc-gh-jcieslak** [comment]
Hey @nicolamarangoni. We have released the fix as part of the [v0.87.0](https://github.com/Snowflake-Labs/terraform-provider-snowflake/releases/tag/v0.87.0) release (but our latest version is [v0.87.2](https://github.com/Snowflake-Labs/terraform-provider-snowflake/releases/tag/v0.87.2)). Please follow the [migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/v0.87.0/MIGRATION_GUIDE.md#v0860--v0870) during the update. Please confirm that the issue is resolved in the newest version. Thanks!

**Comment by nicolamarangoni** [comment]
Hello @sfc-gh-jcieslak, since version 0.87 I cannot create databases from shares anymore:
```
╷
│ Error: error creating database SHARE_JOYN: [errors.go:17] object does not exist or not authorized
│ 
│   with snowflake_database.from_share["myshare"],
│   on snowflake_shares.tf line 1, in resource "snowflake_database" "from_share":
│    1: resource "snowflake_database" "from_share" {
```
This works until 0.86.
The role used is ACCOUNTADMIN. This is set in the provider-setting and also in the in the env variable SNOWFLAKE_ROLE.

**Comment by nicolamarangoni** [comment]
@sfc-gh-jcieslak I checked the query history and you have a bug in the code generation.
The generated sql in version 0.87.x is:
`CREATE DATABASE "MYSHARE" FROM SHARE "NC34666.SHAREDDATA"."SHARED_SCHEMA";`
If I change `"NC34666.SHARED"` to `"NC34666"."SHAREDDATA"` the statement is correct and works!

**Comment by sfc-gh-jcieslak** [comment]
Hey @nicolamarangoni 
That's strange, but even if possibly something changed, this configuration should work for you
```terraform
resource "snowflake_database" "..." {
   from_share = {
      provider = "NC34666"
      share = "SHAREDDATA"
   }
}
```

**Comment by nicolamarangoni** [comment]
Hey @sfc-gh-jcieslak I tried that but it doesn't help.

**Comment by sfc-gh-jcieslak** [comment]
Hmm, what is logged then? If you're getting this Snowflake error (`[errors.go:17] object does not exist or not authorized`), then it's indicating you are probably referencing something that is not created or you don't have sufficient privileges to see that resource. If that's the case, please take a look [here](https://docs.snowflake.com/en/user-guide/data-sharing-multiple-db) to have a checklist of needed things to create a database from share. Also, with the change I proposed, SQL was built properly (could you paste it)?

**Comment by nicolamarangoni** [comment]
@sfc-gh-jcieslak this is the creation statement that is created by version 0.87 of the provider:
`CREATE DATABASE "MYSHARE" FROM SHARE "PROVIDER.SHARE"."SHARED_SCHEMA";`
It should be:
`CREATE DATABASE "MYSHARE" FROM SHARE "PROVIDER"."SHARE"."SHARED_SCHEMA";`

**Comment by sfc-gh-jcieslak** [comment]
Here's an example configuration for sharing that works
```terraform
# Account 1
resource "snowflake_share" "test" {
  provider = snowflake.sharing # provider configuration that connects to secondary account
  name = "sharing_test_share"
  accounts = [ "<org-name>.<acc-name>" ] # place where you want to share
  # org-name can be retrieved with: select current_organization_name();
  # acc-name can be retrieved with: select current_account_name();
}

resource "snowflake_database" "test" {
  provider = snowflake.sharing
  name = "sharing_test_database"
}

resource "snowflake_grant_privileges_to_share" "test" {
  provider = snowflake.sharing
  privileges = [ "USAGE" ]
  on_database = snowflake_database.test.name
  to_share = snowflake_share.test.name 
}

# Account 2
resource "snowflake_database" "database_from_share" {
  name = "example_database_from_share"
  from_share = {
    provider = "<account_locator>" # account locator of the sharing account
    # account_locator can be retrieved with: select current_account()
    share = snowflake_share.test.name
  }
}
```
I think you have to adjust your provider to use account_locator instead of organization + account name

**Comment by nicolamarangoni** [comment]
I think that there is something weird with a changed behavior in Snowflake itself.
In my setup I used
```
provider = "nc34666.joyn_share"
share = "p7s1_share"
```
this worked until 0.86, starting from 0.87 it doesn't wok anymore.
Possibly because "provider" now is only the locator.
with `SHOW SHARES` i get:
owner_account = NC34666.JOYN_SHARE
name = P7S1_SHARE
If I try using as provider the locator alone, I get the same error.

The only thing that works in my setup is to create the db from an identifier with 3 component:
locator.something.share.
I will contact the Snowflake support for an explanation.

**Comment by sfc-gh-jcieslak** [comment]
Yes, that's what It said in the example. You should put account locator to the `provider` field and not `orgname` + `account name` like right now. Call `select current_account()` on the sharing account and you'll get the value you need to provide to the `provider` field.

**Comment by nicolamarangoni** [comment]
Hi @sfc-gh-jcieslak ,
there were some misunderstandings with the issue because the account which contain the share belongs to another department within our organization.
`nc34666` is actually the organization not the locator, I thought only locators are made of 2 letters and 5 numbers. 
`joyn_share` is actually the account name.
The actual locator is `bc71502`, with this locator I could create the database on the share.
Basically the newest version of the Terraform provider doesn't allow to create shares with `organization.accountname` as provider anymore.

**Comment by sfc-gh-jcieslak** [comment]
Yes, after the update the input of `provider` is properly quoted. Without previous quoting things like `organization.accountname` formats were accepted, but was happened rather "by accident". Please use the account locator format for now. In the future we'll be migrating it to the `organization.accountname` format as it's recommended by Snowflake. In the change introducing such change, we'll add a note about in the [migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/MIGRATION_GUIDE.md).

**Comment by nicolamarangoni** [comment]
In any case the issue with the retention is not yet solved.
If the database is not existing and I define it without data_retention_time_in_days, it is successfully created with retention -1:
```terraform
  # snowflake_database.from_share["my_share"] will be created
  + resource "snowflake_database" "from_share" {
      + data_retention_time_in_days = -1
      + from_share                  = {
          + "provider" = "MY_LOCATOR"
          + "share"    = "MY_SHARE"
        }
      + id                          = (known after apply)
      + is_transient                = false
      + name                        = "my_share"
    }
```
Every successive deployment want to set data_retention_time_in_days = 0:
```terraform
  # snowflake_database.from_share["my_share"] will be updated in-place
  ~ resource "snowflake_database" "from_share" {
      ~ data_retention_time_in_days = 0 -> -1
        id                          = "MY_SHARE"
        name                        = "MY_SHARE"
        # (2 unchanged attributes hidden)
    }
```
For the moment, I need to keep `data_retention_time_in_days = 0`

**Comment by nicolamarangoni** [comment]
DATA_RETENTION_IN_DAYS = 1 in the account.
the result of `show parameters for database SHARE_JOYN;` doesn't contain the parameter DATA_RETENTION_IN_DAYS.
For Terraform, the database from share has DATA_RETENTION_IN_DAYS = 0.

**Comment by sfc-gh-jcieslak** [comment]
`show databases` returns data retention param for databases in the dedicated column. It seems like a normal Snowflake behavior. When you create a database like so:
```sql
create database test_database;
```
then call `show databases like 'test_database';` it will return retention_time equal to the parameter default value, but when creating a database from share:
```sql
create database shared_database from share xxx;
```
`show databases like 'shared_database';` will return 0, so your diff changes suggested by the provider are correct.

**Comment by sfc-gh-jcieslak** [comment]
Hey @nicolamarangoni 👋 
Are we good to close the issue or there's something else needed here ?

**Comment by nicolamarangoni** [comment]
Hi @sfc-gh-jcieslak , I will close this issue.
Thanks!
